### PR TITLE
fix: Discard invalid orphan blocks

### PR DESCRIPTION
**Pull request by keroro520**
* Discard invalid orphan blocks
    When accepts a new block, its descendants should be accepted too if valid. So if an error occurs when we try to accept its descendants, the descendants are invalid.

### CODE REVIEW

File: sync/src/synchronizer/mod.rs [hunk 1]
```diff
@@ -257,25 +257,26 @@ impl<CS: ChainStore> Synchronizer<CS> {
                     .remove_blocks_by_parent(&block.header().hash());
                 for block in pre_orphan_block {
                     let block = Arc::new(block);
-                    if self
+                    let ret = if self
```
Comment (quake) [inline sync/src/synchronizer/mod.rs]: we can use rust idiomatic for option and result, it's easier to read:
```
if let Some(Err(e)) = self.shared.block_header(&block.header().parent_hash())
    .and(|| self.accept_block(peer, &block).err())
    .or(|| Err(err_msg(...))) {
     debug!(...)
}
```

Comment (keroro520) [reply sync/src/synchronizer/mod.rs]: I have tried to change it more rust idiomatic but failed...
Could you give me the complete code snips? Thanks ~

Comment (quake) [reply sync/src/synchronizer/mod.rs]: Could you try this?  my previous comment is something wrong, sorry.
```
                    match self
                        .shared
                        .block_header(&block.header().parent_hash())
                        .and(Some(self.accept_block(peer, &block)))
                    {
                        None => {
                            debug!(
                                target: "sync", "[Synchronizer] accept orphan block {:?}, parent not found",
                                block
                            );
                        }
                        Some(Err(e)) => {
                            debug!(
                                target: "sync", "[Synchronizer] accept orphan block {:?} error {:?}",
                                block,
                                e
                            );
                        }
                        Some(Ok(_)) => {}
                    }
```

Comment (keroro520) [reply sync/src/synchronizer/mod.rs]: Thanks.
I have totally rewritten this function. PTAL

**Pull request merged**